### ISSUE TITLE
Use better build system

### ISSUE THREAD

**Issue by jonls** [freeserf/freeserf#344]
Currently we have the autotools build system and the Visual Studio build files. Both build systems have to be maintained in parallel which is not ideal.

- CMake as been proposed as a solution. Unlike autotools, CMake can generate the Visual Studio project files as well as Makefiles which would be the main benefit with switching to CMake(?). CMake build files were submitted in #304 by @khillah but need to be rebased and submitted as a separate PR. This also needs updates to `.travis.yml` and `appveyor.yml`.
- I did a test of [Meson](http://mesonbuild.com/) which is a newer build system (here: https://github.com/jonls/freeserf/tree/meson-test). Meson has some interesting features and the configuration setup appears to be simpler than CMake. However, I think Meson is currently lacking some features that we need. For example, compiling with the Visual Studio backend in Meson still requires `pkg-config` to find dependencies and there is apparently no way to manually override the location of the dependencies (see https://github.com/mesonbuild/meson/issues/1525). I didn't find a way to make this work on Appveyor. With this experience, I think I'd recommend that we go with a more mature build system.

Compiling and running the tests in the `tests/` directory is one thing that may be hindering the switch to another build system. The tests currently use the TAP protocol which is supported by autotools but may be more difficult to get working on other build systems. When testing Meson I ported the tests to use the Google C++ test framework (see https://github.com/jonls/freeserf/commit/54696c301d0a7720d5e7f923ad63a47a0b5a532c) which I think is more appropriate than writing TAP tests directly. I think this switch should also be done when switching the build system.

**Comment by wdigger** [comment]
CMake have CTest and CPack. I try to migrate project to all this systems but don't finish this work.
https://github.com/wdigger/freeserf/tree/cmake

**Comment by jonls** [comment]
@wdigger What is currently missing to finish your CMake work?

CTest seems to provide a way of running tests but a test framework (or TAP runner if the current tests are unchanged) would be needed. Is this correct?

**Comment by wdigger** [comment]
@jonls I did not test the build and packaging on linux. I'll try to do it in the near future.

Mostly popular ctest + gtest variant.

**Comment by wdigger** [comment]
I done porting to CMake.
https://github.com/wdigger/freeserf/tree/cmake

**Comment by wdigger** [comment]
I propose switch to the gtest in any case, make it as a separate task.

**Comment by jonls** [comment]
@wdigger Great, are you opening a PR? It seems like the tests are current not run on Travis, only `make package`. Is this correct? Can you adapt the CMake build scripts to pull in gtest (in another PR?)

**Comment by wdigger** [comment]
@jonls PR created, test added to travis build.
I rebase you gtest changes to cmake, if you do not mind.

**Comment by wdigger** [comment]
@jonls done with gtest
PR after merge CMake to master

**Comment by jonls** [comment]
@wdigger 👍  Closing this issue.

**Comment by khillah** [comment]
Whoa, nice to see this project is still alive :-)
I think I will looka around and consider further work on this code. :+1:
